memiavl/db.go

```diff
@@ -0,0 +1,184 @@
+package memiavl
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+// DB implements DB-like functionalities on top of MultiTree:
+// - async snapshot rewriting
+// - Write-ahead-log
+//
+// The memiavl.db directory looks like this:
+// ```
+// current -> snapshot-N
+// snapshot-N
+//
+//	bank
+//	  kvs
+//	  nodes
+//	  metadata
+//	acc
+//	... other stores
+//
+// wal
+// ```
+type DB struct {
+	MultiTree
+	dir string
+
+	snapshotRewriteChan   chan snapshotResult
+	snapshotRewriteBuffer []MultiChangeSet
+}
+
+type Options struct {
+	CreateIfMissing bool
+	InitialVersion  uint32
+	// the initial stores when initialize the empty instance
+	InitialStores []string
+}
+
+func Load(dir string, opts Options) (*DB, error) {
+	currentDir := currentPath(dir)
+	mtree, err := LoadMultiTree(currentDir, opts.InitialVersion)
+	if err != nil {
+		if opts.CreateIfMissing && os.IsNotExist(err) {
+			tmp := NewEmptyMultiTree(opts.InitialStores, 0)
+			snapshotDir := snapshotPath(dir, 0)
+			if err := tmp.WriteSnapshot(snapshotDir); err != nil {
+				return nil, err
+			}
+			if err := os.Symlink(snapshotDir, currentDir); err != nil {
+				return nil, err
+			}
+			mtree, err = LoadMultiTree(currentDir, opts.InitialVersion)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &DB{
+		MultiTree: *mtree,
+		dir:       dir,
+	}, nil
+}
+
+// Commit wraps `MultiTree.ApplyChangeSet` to add some db level operations:
+// - manage background snapshot rewriting
+// - write WAL
+func (t *DB) Commit(changeSets MultiChangeSet) ([]byte, int64, error) {
+	if t.snapshotRewriteChan != nil {
+		// check the completeness of background snapshot rewriting
+		select {
+		case result := <-t.snapshotRewriteChan:
+			rewriteBuffer := t.snapshotRewriteBuffer
+			t.snapshotRewriteChan = nil
+			t.snapshotRewriteBuffer = nil
+
+			if result.mtree == nil {
+				// background snapshot rewrite failed
+				return nil, 0, fmt.Errorf("background snapshot rewriting failed: %w", result.err)
+			}
+			// snapshot rewrite succeeded, switch and catchup
+			// TODO replay buffer in background to minimize blocking on main thread
+			t.MultiTree = *result.mtree
+			for _, cs := range rewriteBuffer {
+				if _, _, err := t.ApplyChangeSet(cs, false); err != nil {
+					return nil, 0, fmt.Errorf("snapshot rewrite buffer replay failed: %w", err)
```

Comment:
to avoid block main thread
```suggestion
			chErr := make(chan error)
			go func() {
				for _, cs := range rewriteBuffer {
					if _, _, err := t.ApplyChangeSet(cs, false); err != nil {
						chErr <- fmt.Errorf("snapshot rewrite buffer replay failed: %w", err)
						return
					}
				}
				chErr <- nil
			}()
			if err := <-chErr; err != nil {
				return nil, 0, err
```

Reply:
isn't it block on the `chErr` instead? I think it's not so trivial ;D

Reply:
and `t.ApplyChangeSet` can't be called concurrently.

Reply:
yea, it's only put to goroutine in order

Reply:
will solve in a separate PR.